internal_packages/keybase/lib/decrypt-button.cjsx

```diff
@@ -29,80 +32,112 @@ class DecryptMessageButton extends React.Component
+  decryptAttachmentsPopoverDone: (passphrase) =>
+    for recipient in @props.message.to
       privateKeys = PGPKeyStore.privKeys(address: recipient.email, timed: false)
       for privateKey in privateKeys
         PGPKeyStore.getKeyContents(key: privateKey, passphrase: passphrase, callback: (identity) => PGPKeyStore.decryptAttachments(identity, @state.encryptedAttachments))
 
+  _openPassphrasePopover: (target, callback) =>
+    Actions.openPopover(
+        <PassphrasePopover addresses={_.pluck(@props.message.to, "email")} onPopoverDone={callback} />,
+        {originRect: target, direction: 'down'}
+      )
+
+  _noPrivateKeys: =>
+    numKeys = 0
+    for recipient in @props.message.to
+      numKeys = numKeys + PGPKeyStore.privKeys(address: recipient.email, timed: false).length
+    return numKeys < 1
+
   render: =>
-    # TODO inform user of errors/etc. instead of failing without showing it
     if not (@state.wasEncrypted or @state.encryptedAttachments.length > 0)
       return false
 
-    # TODO a message saying "this was decrypted with the key for ___@___.com"
-    title = if @state.isDecrypted then "Message Decrypted" else "Message Encrypted"
+    title = "Message Encrypted"
+    decryptLabel = "Decrypt"
+    borderClass = "border"
+    decryptClass = "decrypt-bar"
+    if @state.status?
+      console.warn(@state.status)
+      if @state.status.indexOf("Message decrypted") >= 0
+        title = @state.status
+        borderClass = "border done-border"
+        decryptClass = "decrypt-bar done-decrypt-bar"
+      else if @state.status.indexOf("Unable to decrypt message.") >= 0
+        title = @state.status
+        borderClass = "border error-border"
+        decryptClass = "decrypt-bar error-decrypt-bar"
+        decryptLabel = "Try Again"
 
     decryptBody = false
-    if !@state.isDecrypted
-      decryptBody = <button title="Decrypt email body" className="btn btn-toolbar" onClick={@_onClickDecrypt} ref="button">Decrypt</button>
+    if !@state.isDecrypted and !(@state.status?.indexOf("malformed") >= 0)
+      decryptBody = <button title="Decrypt email body" className="btn btn-toolbar" onClick={@_onClickDecrypt} ref="button">{decryptLabel}</button>
 
     decryptAttachments = false
-    if @state.encryptedAttachments?.length == 1
-      decryptAttachments = <button onClick={ @_onClickDecryptAttachments } className="btn btn-toolbar">Decrypt Attachment</button>
-      title = "Attachment Encrypted"
-    else if @state.encryptedAttachments?.length > 1
-      decryptAttachments = <button onClick={ @_onClickDecryptAttachments } className="btn btn-toolbar">Decrypt Attachments</button>
-      title = "Attachments Encrypted"
-
+    if @state.encryptedAttachments?.length >= 1
+      title = if @state.encryptedAttachments?.length == 1 then "Attachment Encrypted" else "Attachments Encrypted"
+      buttonLabel = if @state.encryptedAttachments?.length == 1 then "Decrypt Attachment" else "Decrypt Attachments"
```

Comment:
This will look great once we're able to convert this to ES6 ;-) and get some ternaries going on.